scipy/sparse/linalg/isolve/lsmr.py

```diff
@@ -221,6 +221,8 @@ def lsmr(A, b, damp=0.0, atol=1e-6, btol=1e-6, conlim=1e8,
     if maxiter is None:
         maxiter = minDim
 
+    dtype = result_type(b, float)
```

Comment:
dtype of A probably also should be checked?

Does LSMR work for complex A?
If yes, the docstring should be updated for the meaning of `^T` (does it become `^H`?).

Reply:
True, and actually one of the tests already covered that case. I've extended the dtype to also consider `x0` if given.

More tests included now.